Implement filtering by workflow state

ashklianko opened the pull request.

ashklianko committed:
Implement filtering by workflow state #4468

-Added some missing on frontend filters

ashklianko committed:
Implement filtering by workflow state #4468

-fix of npe

alansemenov reviewed (changes requested):
Duplicating review from the closed PR here:

1. We should make sure that none of the Workflow states are hardcoded. 
For example, `WorkflowState[WorkflowState.READY].toLowerCase()` will give you 'ready' and so on. You can also create an instance of the `Workflow` class and then you can use `getStateAsString()` to get string value.
2. Use existing `status.workflow.in_progress` and `status.workflow.read` literals for texts, no reason for new translations. You can also get correct literal id by concatenating `status.workflow.` with `Workflow.getStateAsString()`

ashklianko committed:
Implement filtering by workflow state #4468

-also Modifier field

ashklianko commented:
Updated, also made a first portion of refactoring. Modifier field added

ashklianko committed:
Implement filtering by workflow state #4468

-codacy

ashklianko committed:
Implement filtering by workflow state #4468

-codacy

ashklianko committed:
Implement filtering by owner #4471

alansemenov reviewed (changes requested):
Last commit is linked to the issue of adding filtering by Owner, but in fact adding filtering by Modifier... It's not the same thing, these are two different properties.

Also, please read the specs of how filters by Owner and Modifier should look like: https://github.com/enonic/app-contentstudio/issues/4466. It should not be aggregation checkboxes like for other fields, it should be a radio group with two radio "Me" (Current user) and a dropdown with aggregations.

To make it a bit easier, we can have the checkboxes but the first one should always be "Me" and represent current user.

Workflow is fine now.

ashklianko committed:
Implement filtering by language #4472

-Moved updating aggregations display names into separate class
-Pushed current user (Me) on top of aggregations

ashklianko commented:
@alansemenov, update on PR:
- All required aggregations added
- Currently new aggregations may have just one selection,  ContentBrowseFilterPanel logic is to be reworked to allow multiple selection, current logic is extensive and requires multiple queries requests, will work on it soon
- After multiple selection for new aggregations will start working I will add aggregation dropdowns as mentioned in issue's epic

alansemenov commented on modules/lib/src/main/resources/assets/js/app/browse/filter/AggregationsProcessor.ts:
```diff
@@ -0,0 +1,141 @@
+
+    private updateWorkflowAggregation(workflowAggr: BucketAggregation, total: number): void {
+        // contents might not have a workflow property, thus aggregation won't see those contents, but they are treated as ready
+        const inProgressKey: string = WorkflowState[WorkflowState.IN_PROGRESS].toLowerCase();
```
`WorkflowState[WorkflowState.IN_PROGRESS].toLowerCase()`

As we seem to use this all over the place and since `Workflow.getStateAsString` is not used at the moment, I suggest we change it like this:

```Typescript
    static getStateAsString(state: WorkflowState): string {
        return WorkflowState[state].toLowerCase();
    }
```

and start using `Workflow.getStateAsString(WorkflowState.IN_PROGRESS)`

ashklianko replied:
Or since we're going to modify lib-admin-ui then we can do this:
```
export enum WorkflowState {
    READY = 'ready', IN_PROGRESS = 'in_progress'
}
```

and remove `getStateAsString` at all

alansemenov replied:
Yeah, this is even better. I assume we are not sending numeric enum values to the API.

ashklianko committed:
Search improvements #4466

-multiselection for new aggregations

Commit:
ui-tests issue-4518

ashklianko committed:
Search improvements #4466

-maintaining order of aggregations

ashklianko committed:
Search improvements #4466

-lint issues
-fixed workflow total count
-code ref on filter panel

ashklianko committed:
Search improvements #4466

-removed redundant code

Commit:
ui-tests issue#4512

ashklianko committed:
Search improvements #4466

-refactoring required for studio plus

ashklianko committed:
Search improvements #4466

-Changes to make required methods to be overriden

ashklianko committed:
Search improvements #4466

-allow reuse date filters

Commit:
ui-test issue#4508

ashklianko committed:
Replace checkboxes with dropdowns for Owner and Modifier #4538

-Added new filterable buckets views, allowing to keep all buckets in a dropdown, and show them on the top when selected. Certain items may be always kept on top no matter selected or not

ashklianko committed:
lint

ashklianko committed:
Need to remove selected bucket view from top when deselected

ashklianko committed:
Implement filtering by workflow state #4503

-moving Workflow from lib-admin to the Content Studio

ashklianko committed:
Implement filtering by workflow state #4503

-update all

ashklianko committed:
-fix for build

alansemenov committed:
Refactoring (#4468)

alansemenov reviewed (changes requested):
<img width="240" alt="image" src="https://user-images.githubusercontent.com/11612490/168922067-6477d325-c80a-4e16-be8f-ff1fb616488d.png">

1. Current user or Me should not be in the user selector as we already have it above. Having it in both places is confusing.
2. If there are no users inside the selector (for example, current user is owner of all items), then selector should not be shown, only the "Me" checkbox.
3. Font inside the selector should be the same as for the checkboxes above the selector.
4. OK button should have the blue background instead of green.

ashklianko commented:
> <img alt="image" width="240" src="https://user-images.githubusercontent.com/11612490/168922067-6477d325-c80a-4e16-be8f-ff1fb616488d.png">
> 
> 1. Current user or Me should not be in the user selector as we already have it above. Having it in both places is confusing.
> 2. If there are no users inside the selector (for example, current user is owner of all items), then selector should not be shown, only the "Me" checkbox.
> 3. Font inside the selector should be the same as for the checkboxes above the selector.
> 4. OK button should have the blue background instead of green.

Ok for all points except 1) - it is much more handy to let user deselect multiple items at once, including Me, as in the list above, and it is also consistent with other options selected and you can select/deselect them in both list above and selector

ashklianko committed:
Implement filtering by workflow state #4468

-Changed font size for buckets in the dropdown list to be same as in aggregation on top
-Hiding buckets selector dropdown when it contains same items as those that to be always shown on top

ashklianko committed:
Implement filtering by workflow state #4468

-lint fixes

ashklianko committed:
Implement filtering by workflow state #4468

-styling dropdown buckets same as regular ones

Commit:
skip failed test

Commit:
ui-tests for aggregations

alansemenov committed:
Modifier -> Last modified by (#4468)

ashklianko committed:
Implement filtering by workflow state #4468

-refreshing filter's aggregations on content update when filter is open/opened or if grid is filtered

Commit:
update failed ui-test
